staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go

```diff
@@ -308,3 +317,37 @@ func (tw *baseTimeoutWriter) Hijack() (net.Conn, *bufio.ReadWriter, error) {
 	}
 	return conn, rw, err
 }
+
+// timeoutReader wraps a http.Request.Body (an io.ReadCloser) to make
+// read calls fail once the timeout has occurred.
+type timeoutReader struct {
+	lock sync.Mutex
+	io.ReadCloser
+	timedOut bool
+}
+
+func (tr *timeoutReader) timeout() {
+	tr.lock.Lock()
+	defer tr.lock.Unlock()
+	tr.timedOut = true
+}
+
+func (tr *timeoutReader) Read(b []byte) (int, error) {
+	tr.lock.Lock()
+	defer tr.lock.Unlock()
+
+	if tr.timedOut {
+		return 0, http.ErrHandlerTimeout
+	}
+	return tr.ReadCloser.Read(b)
```

Comment:
Suppose `Read` blocks because it is waiting for the HTTP client to send the body. Then the lock is held and `timeout()` in `timeoutHandler.ServeHTTP` blocks, preventing the timeout handling.

Reply:
yes, this could happen. I think we already have this today for the `ResponseWriter`
https://github.com/kubernetes/kubernetes/blob/ad18954259eae3db51bac2274ed4ca7304b923c4/staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go#L149

`ioutil.ReadAll` reads 512 bytes at a time, so the the `Read` calls will probably be intermingled between the two goroutinesin in this situation:
- goroutine `B` does a `Read` and reads some bytes from the Request Body (before timeout)
- request times out
- timeout handler marks the body reader as `timed out`
- goroutine `B` calls `Read` and gets `ErrHandlerTimeout`

I don't think we need to prevent `Read` calls getting intermingled, the request has timed out and we are sending a `504` to the caller.

but if a `Read` or `Write` call blocks from goroutine `B` it will freeze the timeout handler and prevent it from sending a panic or 504 indefinitely.